### PR TITLE
Neutral Dixon Coles model

### PR DESCRIPTION
**Pull request by jack89roberts**
Dixon Coles model modified to allow for:
- Matches at neutral venues (e.g. international tournaments)
- Separate attack and defence, home and away, (dis)advantages for each team

TODO - reduce duplicated code between `ExtendedDixonColes` and `NeutralDixonColes`

### CODE REVIEW

File: bpl/neutral_dixon_coles.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+    # pylint: disable=arguments-differ,too-many-arguments
+    def fit(
+        self,
+        training_data: Dict[str, Union[Iterable[str], Iterable[float]]],
+        random_state: int = 42,
+        num_warmup: int = 500,
+        num_samples: int = 1000,
+        mcmc_kwargs: Optional[Dict[str, Any]] = None,
+        run_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> ExtendedDixonColesMatchPredictor:
```
Comment (rchan26) [inline bpl/neutral_dixon_coles.py]: should this be `NeutralDixonColesMatchPredictor`?

File: bpl/neutral_dixon_coles.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+                "standardised_defence",
+                dist.Normal(
+                    loc=rho * standardised_attack, scale=jnp.sqrt(1.0 - rho**2.0)
+                ),
+            )
+
+            with reparam(config={"home_attack_advantage": LocScaleReparam(centered=0)}):
+                home_attack_advantage = numpyro.sample(
+                    "home_attack_advantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
```
Comment (rchan26) [inline bpl/neutral_dixon_coles.py]: Shouldn't you rename these parameters to something like `mean_home_attack_adv` and `std_home_attack_adv`

File: bpl/neutral_dixon_coles.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+                )
+            with reparam(
+                config={"away_attack_disadvantage": LocScaleReparam(centered=0)}
+            ):
+                away_attack_disadvantage = numpyro.sample(
+                    "away_attack_disadvantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
```
Comment (rchan26) [inline bpl/neutral_dixon_coles.py]: should have their own parameters for the mean and st-dev (i.e `mean_away_attack_disadv` and `std_away_attack_disadv`) and add corresponding priors for them above

I would actually maybe suggest dropping the `_advantage` and `_disadvantage` and then just have the attack and defence variables interpreted like we have already (so a positive attack means good attack, and positive defence means good attack - and make it *minus* the defence parameter) - would help avoid the really long variable names

File: bpl/neutral_dixon_coles.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+                )
+            with reparam(
+                config={"home_defence_advantage": LocScaleReparam(centered=0)}
+            ):
+                home_defence_advantage = numpyro.sample(
+                    "home_defence_advantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
```
Comment (rchan26) [inline bpl/neutral_dixon_coles.py]: see above comments

File: bpl/neutral_dixon_coles.py [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+                )
+            with reparam(
+                config={"away_defence_disadvantage": LocScaleReparam(centered=0)}
+            ):
+                away_defence_disadvantage = numpyro.sample(
+                    "away_defence_disadvantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
```
Comment (rchan26) [inline bpl/neutral_dixon_coles.py]: see above comments

File: bpl/neutral_dixon_coles.py [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+        )
+        defence = numpyro.deterministic(
+            "defence", defence_prior_mean + standardised_defence * std_defence
+        )
+
+        expected_home_goals = jnp.exp(
+            attack[home_team]
+            - defence[away_team]
+            + (1 - neutral_venue) * home_attack_advantage[home_team]
+            + (1 - neutral_venue) * away_defence_disadvantage[away_team]
```
Comment (rchan26) [inline bpl/neutral_dixon_coles.py]: carrying from previous comment - maybe you have:
```
expected_home_goals = jnp.exp(
            attack[home_team]
            - defence[away_team]
            + (1 - neutral_venue) * home_attack[home_team]
            - (1 - neutral_venue) * away_defence[away_team]
```

File: bpl/neutral_dixon_coles.py [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+        expected_away_goals = jnp.exp(
+            attack[away_team]
+            - defence[home_team]
+            - (1 - neutral_venue) * away_attack_disadvantage[away_team]
+            - (1 - neutral_venue) * home_defence_advantage[home_team]
+        )
+
+        # FIXME: this is because the priors allow crazy simulated data before inference
+        expected_home_goals = jnp.clip(expected_home_goals, a_max=15.0)
+        expected_away_goals = jnp.clip(expected_away_goals, a_max=15.0)
```
Comment (rchan26) [inline bpl/neutral_dixon_coles.py]: not 100% sure if this is the right fix - but I think we need to bound the correlation term (like it does in the paper). The following code just places a uniform distribution between the minimum and maximum defined in the paper
```
# impose bounds on the correlation coefficient
corr_coef_raw = numpyro.sample("corr_coef_raw", dist.Uniform(low=0.0, high=1.0))
UB = jnp.min(jnp.array([jnp.min(1.0/(expected_home_goals*expected_away_goals)), 1]))
LB = jnp.max(jnp.array([jnp.max(-1.0/expected_home_goals), jnp.max(-1.0/expected_away_goals)]))
corr_coef = numpyro.deterministic("corr_coef", LB + corr_coef_raw * (UB-LB))
```
and this would mean deleting the original definition of `corr_coef` above.

I've tried this out and deleted the `tol` parameter in the `dixon_coles_correlation_term` and this ran fine - I'm not sure when this error originally came up though...

File: bpl/neutral_dixon_coles.py (hunk 1 above)
Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: yes!

File: bpl/neutral_dixon_coles.py (hunk 7 above)
Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: Not sure about this one, need to dig into it/have a chat at some point to understand. All of this is carried over from the 'normal' BPL model, it's not been changed for the neutral model.

File: bpl/neutral_dixon_coles.py (hunk 3 above)
Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: The idea was meant to be that, with the way they're defined, they can all have the same priors because we expect them all to have around the same mean etc. (e.g. home attack advantage ~= away defence disadvantage). Does that make sense? The names are pretty clunky though.

File: bpl/neutral_dixon_coles.py (hunk 7 above)
Comment (rchan26) [reply bpl/neutral_dixon_coles.py]: I'm happy to ignore this for now - just a thought that could be why you had these as a quick fix. The paper has bounds for this and so thought would be good to implement them too

File: bpl/neutral_dixon_coles.py (hunk 3 above)
Comment (rchan26) [reply bpl/neutral_dixon_coles.py]: I see - this makes sense. I'm not a good enough Bayesian to argue otherwise but I just felt that they needed their own, even if they are the same prior. Maybe just a clean up of the variable names so that it's clear they're meant to be a general prior for these parameters would be good

Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: To be clear I'm asking whether that makes sense because I have no idea really 😄

Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: Just to update this thread - I now agree that we need a separate variable for all of these as we're taking a sample from the prior, rather than defining it. Ryan's done this in #14

File: bpl/neutral_dixon_coles.py (hunk 7 above)
Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: I've had a look back at the paper and see what you're doing here now. Wondering whether a truncated normal would be better than uniform (it would be closer to what we're doing now at least)? We'd also need to update this in all the models, not just the Neutral one (at least if changing dixon_coles_correlation_term).

Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: Truncated normal may have too low probability in the tails possibly, maybe something inbetwteen the two

Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py:157]: (e.g. truncated normal with a larger variance)

Comment (rchan26) [reply bpl/neutral_dixon_coles.py]: We can maybe do a beta(2,2) distribution? It's bounded between 0 and 1 so it can work nicely I think if we wanted something bit more informative than a uniform

I can make that change in my PR now to pull into this branch and make the changes in the standard and extended dixon coles models

Comment (jack89roberts) [reply bpl/neutral_dixon_coles.py]: I was wondering about beta but I think rho can be negative looking at the paper (also in what you did above)

Ah I guess you mean sampling the initial `_raw` one from beta and then doing the same transformation?

Comment (rchan26) [reply bpl/neutral_dixon_coles.py]: Yeah that's what I was doing (should've commented the code better!):

I sample from an initial distribution which is bounded (so this is either uniform or beta), the `_raw` variable, and then transform this using the bounds which gives me a distribution in `[LB, UB]` where they are defined in the paper and I take the minimum or maximum of these values computed using `expected_home_goals` and `expected_away_goals` (which refer to `lambda` and `mu` in the paper respectively).

So currently in the above, `corr_coef` has a uniform distribution in `[LB,UB]`. If we changed this to a Beta(2,2), this would be a scaled Beta(2,2) distribution over `[LB,UB]`, rather than `[0,1]`.

**Comment by jack89roberts** [comment]
To debug: Tests pass locally (on Mac) but not in actions (on Ubuntu)

**Review by nbarlowATI — APPROVED** [review]
Looks good!  (apologies for slow review...).  I think this has been battle-tested enough that we should merge it...   Tests and github action should be working now...
